File: examples/customer_support_agent_ts/src/setup-controls.ts
/**
 * Setup script — creates demo controls, a policy, and assigns it to the agent.
 *
 * Run once after starting the server:
 *   npm run setup
 */

import { AgentControlClient, type ControlDefinitionInput } from "agent-control";

const serverUrl = process.env.AGENT_CONTROL_URL ?? "http://localhost:8000";
const apiKey = process.env.AGENT_CONTROL_API_KEY;
const agentName = "customer-support-agent-ts-2";
const policyName = `policy-${agentName}`;

const client = new AgentControlClient();
client.init({
  agentName,
  serverUrl,
  ...(apiKey ? { apiKey } : {}),
  registerAgent: false,
});

interface ControlSpec {
  name: string;
  definition: ControlDefinitionInput;
}

const DEMO_CONTROLS: ControlSpec[] = [
  {
    name: "block-ssn-in-input",
    definition: {
      description: "Blocks user messages containing SSN patterns",
      enabled: true,
      execution: "server",
      scope: { stages: ["pre"], stepTypes: ["llm"], stepNames: ["respond_to_customer"] },
      selector: { path: "input" },
      evaluator: {
        name: "regex",
        config: { pattern: String.raw`\d{3}-\d{2}-\d{4}` },
      },
      action: { decision: "deny" },
    },
  },
  {
    name: "block-ssn-in-output",
    definition: {
      description: "Blocks LLM responses containing SSN patterns",
      enabled: true,
      execution: "server",
      scope: { stages: ["post"], stepTypes: ["llm"], stepNames: ["respond_to_customer"] },
      selector: { path: "output" },
      evaluator: {
        name: "regex",
        config: { pattern: String.raw`\d{3}-\d{2}-\d{4}` },
      },
      action: { decision: "deny" },
    },
  },
  {
    name: "block-prompt-injection",
    definition: {
      description: "Blocks common prompt injection attempts",
      enabled: true,
      execution: "server",
      scope: { stages: ["pre"], stepTypes: ["llm"], stepNames: ["respond_to_customer"] },
      selector: { path: "input" },
      evaluator: {
        name: "regex",
        config: {
          pattern: String.raw`(?i)(ignore.{0,20}(previous|prior|above).{0,20}instructions|you are now|system:|forget everything|disregard)`,
        },
      },
      action: { decision: "deny" },
    },
  },
  {
    name: "block-credit-card",
    definition: {
      description: "Blocks credit card numbers in input",
      enabled: true,
      execution: "server",
      scope: { stages: ["pre"], stepTypes: ["llm"], stepNames: ["respond_to_customer"] },
      selector: { path: "input" },
      evaluator: {
        name: "regex",
        config: {
          pattern: String.raw`\b\d{4}[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}\b`,
        },
      },
      action: { decision: "deny" },
    },
  },
  {
    name: "block-credit-card-in-output",
    definition: {
      description: "Blocks LLM responses containing credit card numbers",
      enabled: true,
      execution: "server",
      scope: { stages: ["post"], stepTypes: ["llm"], stepNames: ["respond_to_customer"] },
      selector: { path: "output" },
      evaluator: {
        name: "regex",
        config: {
          pattern: String.raw`\b\d{4}[-\s]?\d{4}[-\s]?\d{4}[-\s]?\d{4}\b`,
        },
      },
      action: { decision: "deny" },
    },
  },
  {
    name: "block-sql-injection-lookup",
    definition: {
      description: "Blocks SQL injection in customer lookup tool",
      enabled: true,
      execution: "server",
      scope: {
        stages: ["pre"],
        stepTypes: ["tool"],
        stepNames: ["lookup_customer"],
      },
      selector: { path: "input" },
      evaluator: {
        name: "regex",
        config: {
          pattern: String.raw`(?i)(select|insert|update|delete|drop|union|--|;)`,
        },
      },
      action: { decision: "deny" },
    },
  },
  {
    name: "log-ticket-creation",
    definition: {
      description: "Logs all ticket creation for audit",
      enabled: true,
      execution: "server",
      scope: {
        stages: ["pre"],
        stepTypes: ["tool"],
        stepNames: ["create_ticket"],
      },
      selector: { path: "*" },
      evaluator: {
        name: "regex",
        config: { pattern: ".*" },
      },
      action: { decision: "log" },
    },
  },
];

async function main(): Promise<void> {
  console.log(`Setting up controls on ${serverUrl} ...`);

  const health = await client.system.healthCheck();
  console.log(`Server: ${health.status} (${health.version})`);

  // Register agent (steps are auto-registered when the agent app runs init() — same as Python)
  await client.agents.init({
    agent: {
      agentName,
      agentDescription: "TypeScript customer support agent demo",
    },
  });
  console.log(`Registered agent: ${agentName}`);

  // Create or reuse policy
  let policyId: number;
  try {
    const created = await client.policies.create({ name: policyName });
    policyId = created.policyId;
    console.log(`Created policy: ${policyName} (id=${policyId})`);
  } catch {
    const existing = await client.agents.getPolicy({ agentName });
    policyId = existing.policyId;
    console.log(`Reusing existing policy (id=${policyId})`);
  }

  // Assign policy to agent
  await client.agents.updatePolicy({ agentName, policyId });
  console.log(`Assigned policy to agent`);

  // Create controls and attach to policy
  let created = 0;
  for (const spec of DEMO_CONTROLS) {
    try {
      const result = await client.controls.create({ name: spec.name });
      const controlId = result.controlId;

      await client.controls.updateData({
        controlId,
        body: { data: spec.definition },
      });

      await client.policies.addControl({ policyId, controlId });
      created++;
      console.log(`  + ${spec.name} (id=${controlId})`);
    } catch {
      const list = await client.controls.list({ name: spec.name, limit: 1 });
      if (list.controls.length > 0) {
        const controlId = list.controls[0].id;
        await client.controls.updateData({
          controlId,
          body: { data: spec.definition },
        });
        try {
          await client.policies.addControl({ policyId, controlId });
        } catch {
          // already in policy
        }
        console.log(`  ~ ${spec.name} (id=${controlId}, updated)`);
      }
    }
  }

  console.log(
    `\nDone. ${created} control(s) created, ${DEMO_CONTROLS.length} total configured.`,
  );
}

main().catch((err: unknown) => {
  console.error("Setup failed:", err);
  process.exitCode = 1;
});

File: examples/openshell_incident_responder/setup_controls.py
"""
Setup script for the OpenShell CI triage example.

This example attaches a policy, not direct controls, so the same policy can be
reused across multiple shell-capable agents running in OpenShell sandboxes.
"""

from __future__ import annotations

import asyncio
import os
from datetime import UTC, datetime
from typing import Any

from agent_control import AgentControlClient, agents, controls, policies
from agent_control_models import Agent

AGENT_NAME = "openshell-ci-triage-agent"
POLICY_NAME = "openshell-command-governance"
SERVER_URL = os.getenv("AGENT_CONTROL_URL", "http://localhost:8000")


CONTROL_SPECS: list[tuple[str, dict[str, Any]]] = [
    (
        "deny-secret-file-access",
        {
            "description": "Block shell commands that target secrets or credential-bearing files.",
            "enabled": True,
            "execution": "server",
            "scope": {
                "step_types": ["tool"],
                "step_names": ["execute_shell_command"],
                "stages": ["pre"],
            },
            "condition": {
                "selector": {"path": "input.command"},
                "evaluator": {
                    "name": "regex",
                    "config": {
                        "pattern": (
                            r"(\.env(\.|$))|"
                            r"(id_rsa)|"
                            r"(\.aws/credentials)|"
                            r"(printenv)|"
                            r"((^|\s)(cat|less|more|grep|sed)\s+[^\n]*(secret|token|credential))"
                        )
                    },
                },
            },
            "action": {
                "decision": "deny",
                "message": (
                    "Secret-bearing files and environment material must not be read from the shell. "
                    "Use approved config summaries instead."
                ),
            },
        },
    ),
    (
        "deny-outside-workspace-reads",
        {
            "description": "Block attempts to inspect files outside the demo workspace through the shell.",
            "enabled": True,
            "execution": "server",
            "scope": {
                "step_types": ["tool"],
                "step_names": ["execute_shell_command"],
                "stages": ["pre"],
            },
            "condition": {
                "selector": {"path": "input.command"},
                "evaluator": {
                    "name": "regex",
                    "config": {
                        "pattern": (
                            r"((^|\s)(cat|less|more|head|tail|grep|sed)\s+"
                            r"([^-\s][^\n]*\s+)?/[^\n]*)"
                        )
                    },
                },
            },
            "action": {
                "decision": "deny",
                "message": (
                    "Shell reads must stay inside the current repository workspace. "
                    "Use repository-local files only."
                ),
            },
        },
    ),
    (
        "deny-unsafe-bootstrap-install",
        {
            "description": "Block curl-pipe-shell and similar bootstrap installers.",
            "enabled": True,
            "execution": "server",
            "scope": {
                "step_types": ["tool"],
                "step_names": ["execute_shell_command"],
                "stages": ["pre"],
            },
            "condition": {
                "selector": {"path": "input.command"},
                "evaluator": {
                    "name": "regex",
                    "config": {
                        "pattern": r"(curl[^\n]*\|\s*(bash|sh))|(wget[^\n]*\|\s*(bash|sh))"
                    },
                },
            },
            "action": {
                "decision": "deny",
                "message": (
                    "Bootstrap installers like curl-pipe-shell are not allowed. "
                    "Use approved internal package acquisition paths instead."
                ),
            },
        },
    ),
    (
        "deny-unapproved-package-install",
        {
            "description": "Block generic package installation attempts that are not the approved semgrep flow.",
            "enabled": True,
            "execution": "server",
            "scope": {
                "step_types": ["tool"],
                "step_names": ["execute_shell_command"],
                "stages": ["pre"],
            },
            "condition": {
                "and": [
                    {
                        "selector": {"path": "input.command"},
                        "evaluator": {
                            "name": "regex",
                            "config": {
                                "pattern": r"((^|\s)npm\s+install\b)|((^|\s)pip\s+install\b)"
                            },
                        },
                    },
                    {
                        "not": {
                            "selector": {"path": "input.command"},
                            "evaluator": {
                                "name": "regex",
                                "config": {"pattern": r"\bsemgrep\b"},
                            },
                        }
                    },
                ]
            },
            "action": {
                "decision": "deny",
                "message": (
                    "Generic package installation is not allowed during autonomous triage. "
                    "Only approved tool acquisition paths may be used."
                ),
            },
        },
    ),
    (
        "deny-destructive-shell-operations",
        {
            "description": "Block destructive shell commands even when the binary is allowed in the sandbox.",
            "enabled": True,
            "execution": "server",
            "scope": {
                "step_types": ["tool"],
                "step_names": ["execute_shell_command"],
                "stages": ["pre"],
            },
            "condition": {
                "selector": {"path": "input.command"},
                "evaluator": {
                    "name": "regex",
                    "config": {
                        "pattern": (
                            r"(rm\s+-rf)|"
                            r"(git\s+push[^\n]*--force)|"
                            r"(terraform\s+destroy)|"
                            r"(kubectl\s+delete)|"
                            r"(chmod\s+777)|"
                            r"(shutdown|reboot|mkfs)"
                        )
                    },
                },
            },
            "action": {
                "decision": "deny",
                "message": "Destructive shell operations are not allowed for autonomous remediation.",
            },
        },
    ),
    (
        "steer-unpinned-tool-install",
        {
            "description": "Rewrite package installation to a pinned internal-mirror command.",
            "enabled": True,
            "execution": "server",
            "scope": {
                "step_types": ["tool"],
                "step_names": ["execute_shell_command"],
                "stages": ["pre"],
            },
            "condition": {
                "and": [
                    {
                        "selector": {"path": "input.command"},
                        "evaluator": {
                            "name": "regex",
                            "config": {
                                "pattern": r"((^|\s)uv\s+pip\s+install\b[^\n]*\bsemgrep\b)|((^|\s)pip\s+install\b[^\n]*\bsemgrep\b)"
                            },
                        },
                    },
                    {
                        "not": {
                            "selector": {"path": "input.command"},
                            "evaluator": {
                                "name": "regex",
                                "config": {
                                    "pattern": r"uv\s+pip\s+install\s+--index-url\s+https://packages\.company\.internal/simple\s+semgrep==1\.84\.0"
                                },
                            },
                        }
                    },
                ]
            },
            "action": {
                "decision": "steer",
                "message": "Use pinned tooling from the approved internal mirror.",
                "steering_context": {
                    "message": (
                        '{"rewrite_command": "uv pip install --index-url '
                        'https://packages.company.internal/simple semgrep==1.84.0", '
                        '"reason": "Package installs must use pinned versions from the approved '
                        'internal mirror."}'
                    )
                },
            },
        },
    ),
    (
        "deny-secret-output-leak",
        {
            "description": "Block command output that contains secrets from being surfaced or persisted.",
            "enabled": True,
            "execution": "server",
            "scope": {
                "step_types": ["tool"],
                "step_names": ["execute_shell_command"],
                "stages": ["post"],
            },
            "condition": {
                "selector": {"path": "output.stdout"},
                "evaluator": {
                    "name": "regex",
                    "config": {
                        "pattern": r"(sk-prod-[A-Za-z0-9_-]+)|(BEGIN [A-Z ]*PRIVATE KEY)|(INTERNAL_API_TOKEN=)"
                    },
                },
            },
            "action": {
                "decision": "deny",
                "message": "Command output contains secret material and cannot be returned to the agent.",
            },
        },
    ),
]


async def ensure_agent(client: AgentControlClient) -> None:
    """Register the demo agent."""
    agent = Agent(
        agent_name=AGENT_NAME,
        agent_description=(
            "Daily CI triage agent running inside an OpenShell sandbox "
            "with Agent Control command governance"
        ),
        agent_created_at=datetime.now(UTC).isoformat(),
    )
    await agents.register_agent(client, agent, steps=[])


async def ensure_control(
    client: AgentControlClient,
    name: str,
    definition: dict[str, Any],
) -> int:
    """Create or update a control and return its identifier."""
    try:
        result = await controls.create_control(client, name=name, data=definition)
        return int(result["control_id"])
    except Exception as exc:
        if "409" not in str(exc):
            raise

        listing = await controls.list_controls(client, name=name, limit=1)
        existing = listing.get("controls") or []
        if not existing:
            raise
        control_id = int(existing[0]["id"])
        await controls.set_control_data(client, control_id, definition)
        return control_id


async def ensure_policy(client: AgentControlClient, name: str) -> int:
    """Create or reuse a policy."""
    try:
        result = await policies.create_policy(client, name)
        return int(result["policy_id"])
    except Exception as exc:
        if "409" not in str(exc):
            raise

        response = await agents.get_agent_policies(client, AGENT_NAME)
        policy_ids = response.get("policy_ids") or []
        if policy_ids:
            return int(policy_ids[0])
        raise RuntimeError(
            f"Policy '{name}' already exists but is not attached to agent '{AGENT_NAME}'. "
            "Attach it once in the UI or remove the existing policy before rerunning setup."
        )


async def main() -> None:
    """Create policy-backed controls and attach the policy to the agent."""
    async with AgentControlClient(base_url=SERVER_URL) as client:
        await ensure_agent(client)
        print(f"Registered agent: {AGENT_NAME}")

        policy_id = await ensure_policy(client, POLICY_NAME)
        print(f"Using policy: {POLICY_NAME} (id={policy_id})")

        control_ids: list[int] = []
        for name, definition in CONTROL_SPECS:
            control_id = await ensure_control(client, name, definition)
            await policies.add_control_to_policy(client, policy_id, control_id)
            control_ids.append(control_id)
            print(f"Bound control '{name}' to policy (id={control_id})")

        await agents.add_agent_policy(client, AGENT_NAME, policy_id)
        print(f"Attached policy '{POLICY_NAME}' to agent '{AGENT_NAME}'")

        print()
        print("Setup complete.")
        print("Controls in policy:")
        for control_id, (name, _) in zip(control_ids, CONTROL_SPECS, strict=False):
            print(f"  - {name} (id={control_id})")


if __name__ == "__main__":
    asyncio.run(main())
